modules/packages/Python.chpl

```diff
@@ -0,0 +1,1417 @@
+        // but by setting executable we can reuse the python logic to determine
+        // the locale (in the string sense, not the chapel sense)
+        const executable = string.createBorrowingBuffer(venv) + "/bin/python";
+        checkPyStatus(
+          PyConfig_SetString(
+            cfgPtr, c_ptrTo(config_.executable), executable.c_wstr()));
+      }
+
+      // initialize
+      checkPyStatus(Py_InitializeFromConfig(cfgPtr));
+
+      // add the current directory to the python path
+      var sys = PyImport_ImportModule("sys");
+      this.checkException();
+      var path = PyObject_GetAttrString(sys, "path");
+      this.checkException();
+      PyList_Insert(path, 0, Py_BuildValue("s", "."));
+      this.checkException();
+      Py_DECREF(path);
+      Py_DECREF(sys);
+    }
+    @chpldoc.nodoc
+    proc deinit() {
+      Py_Finalize();
+    }
+    /*
+      Run a string of python code.
+
+      This function will just run the code, it cannot be passed arguments or
+      return values.
+    */
+    proc run(code: string) throws {
+      PyRun_SimpleString(code.c_str());
+      this.checkException();
+    }
+
+    /*
+      Register a custom :type:`TypeConverter` with the interpreter. This allows
+      arbitrary Chapel types to be serialized/deserialized to and from Python
+      objects.
+
+      The registered converter will take precedence over the default converters,
+      allowing users to override the default behavior.
+    */
+    proc registerConverter(in cvt: owned TypeConverter) {
+      converters.pushBack(cvt);
+    }
+
+    /*
+      Compile a lambda function like 'lambda x,: x + 1'
+
+      Note: this only works with lambdas that a tuple of arguments, like 'x,' or '*args'
+    */
+    @chpldoc.nodoc
+    proc compileLambda(l: string): owned PyObject throws {
+      var globals = chpl_PyEval_GetFrameGlobals();
+      // create a globals if it doesn't exist
+      if !globals {
+        globals = Py_BuildValue("{}"); // TODO: need to decrement this?
+      }
+      // this is the equivalent of `eval(compile(l, '<string>', 'eval'))`
+      // we can also do `Py_CompileString` -> `PyFunction_New`?
+      var code = PyRun_String(l.c_str(), Py_eval_input, globals, nil);
+      this.checkException();
+      var obj = new PyObject(this, code);
+      return obj;
+    }
+
+    /*
+      Query to see if an Python exception has occurred. If there is an
+      exception, the Python exception will be thrown as a :type:`Exception`.
+
+      .. note::
+
+         This method should be called after any Python API call that may fail.
+         The wrapping types in this module will call this method automatically,
+         most users should not need to call this method directly.
+    */
+    inline proc checkException() throws {
+      var ptype, pvalue, ptraceback: c_ptr(void);
+      PyErr_Fetch(c_ptrTo(ptype), c_ptrTo(pvalue), c_ptrTo(ptraceback));
+
+      if pvalue {
+        var pvalue_str = PyObject_Str(pvalue);
```

Comment:
It would help me (and anyone reading this code) to define type aliases for `void*` corresponding to different types. E.g.,

```
type PyObjectPtr = c_ptr(void);
```

And explicitly annotate types here and in the prototype declarations of these functions. This way, we have at least a notional idea of what the types are.

Reply:
I think even better, its probably possible to just have the literal opaque extern type of `extern type PyObject`. I will look into that